# HG changeset patch
# User Torsten <mttl@mailbox.org>
# Date 1544734412 -3600
#      Thu Dec 13 21:53:32 2018 +0100
# Node ID 2355f66cf91d1a368cf0632042beafb7255f7a8e
# Parent  3a90eff58fb5d7d85c78f8f9cda8163c60814994
allow to select the gui style in the preferences dialog

* gui-preferences.h: add new constants for key and default of new preference

* main-window.cc (main_window): initialize new class variable for the qt app,
  get name of default style;
  (notice_settings): get desired style from preferences or take the default
  one if the desired one is not found;
  (octave_qt_app::octave_qt_app): do not set windows style by default

* main-window.h: new class variables in main_window for qt app and the name
  of the default style, new public method for getting the qt app in
  octave_qt_app

* settings-dialog.cc (settings_dialog): fill new combo box for style selection
  with available styles and select the one from the settings file or the
  default one;
  (write_changed_settings): store the style name from the new combo box

* settings-dialog.ui: add new combo box for selecting the gui style

diff --git a/libgui/src/gui-preferences.h b/libgui/src/gui-preferences.h
--- a/libgui/src/gui-preferences.h
+++ b/libgui/src/gui-preferences.h
@@ -58,16 +58,19 @@ const QStyle::PixelMetric global_icon_si
 {
   QStyle::PM_SmallIconSize,
   QStyle::PM_ToolBarIconSize,
   QStyle::PM_LargeIconSize
 };
 const gui_pref global_icon_size ("toolbar_icon_size", QVariant (0));
 const gui_pref global_icon_theme ("use_system_icon_theme", QVariant (true));
 
+// Style
+const gui_pref global_style ("style", QVariant ("default"));
+
 // Console preferences
 
 const gui_pref cs_font ("terminal/fontName", QVariant ());
 
 
 // Variable Editor preferences
 
 const gui_pref ve_font ("variable_editor/font_size", QVariant ());
diff --git a/libgui/src/main-window.cc b/libgui/src/main-window.cc
--- a/libgui/src/main-window.cc
+++ b/libgui/src/main-window.cc
@@ -22,16 +22,17 @@ along with Octave; see the file COPYING.
 */
 
 #if defined (HAVE_CONFIG_H)
 #  include "config.h"
 #endif
 
 #include <QKeySequence>
 #include <QApplication>
+#include <QStyleFactory>
 #include <QInputDialog>
 #include <QLabel>
 #include <QMenuBar>
 #include <QMenu>
 #include <QAction>
 #include <QSettings>
 #include <QStyle>
 #include <QStyleFactory>
@@ -155,17 +156,18 @@ namespace octave
     m_app_context.delete_interpreter ();
 
     emit octave_finished_signal (exit_status);
   }
 
   main_window::main_window (octave_qt_app& oct_qt_app,
                             octave_qt_link *oct_qt_lnk)
     : QMainWindow (),
-      m_octave_qt_link (oct_qt_lnk), m_workspace_model (nullptr),
+      m_qt_app (oct_qt_app.qt_app ()), m_octave_qt_link (oct_qt_lnk),
+      m_workspace_model (nullptr),
       m_status_bar (nullptr), m_command_window (nullptr),
       m_history_window (nullptr), m_file_browser_window (nullptr),
       m_doc_browser_window (nullptr), m_editor_window (nullptr),
       m_workspace_window (nullptr), m_variable_editor_window (nullptr),
       m_external_editor (new external_editor_interface (this)),
       m_active_editor (m_external_editor),
       m_settings_dlg (nullptr), m_find_files_dlg (nullptr),
       m_release_notes_window (nullptr), m_community_news_window (nullptr),
@@ -221,16 +223,18 @@ namespace octave
     m_workspace_window = new workspace_view (this);
 
     m_active_editor = m_editor_window;
 
 #if defined (HAVE_QGUIAPPLICATION_SETDESKTOPFILENAME)
     QGuiApplication::setDesktopFileName ("org.octave.Octave.desktop");
 #endif
 
+    m_default_style = m_qt_app->style ()->objectName ();
+
     QSettings *settings = resource_manager::get_settings ();
 
     bool connect_to_web = true;
     QDateTime last_checked;
     int serial = 0;
     m_active_dock = nullptr;
 
     if (settings)
@@ -710,16 +714,28 @@ namespace octave
     QMessageBox::about (this, tr ("About Octave"),
                         QString::fromStdString (message));
   }
 
   void main_window::notice_settings (const QSettings *settings)
   {
     // QSettings pointer is checked before emitting.
 
+    // Get desired style from preferences or take the default one if
+    // the desired one is not found
+    QString preferred_style
+          = settings->value (global_style.key, global_style.def).toString ();
+
+    if (preferred_style == global_style.def.toString ())
+      preferred_style = m_default_style;
+
+    QStyle *new_style = QStyleFactory::create (preferred_style);
+    if (new_style)
+      m_qt_app->setStyle (new_style);
+
     // the widget's icons (when floating)
     QString icon_set
       = settings->value ("DockWidgets/widget_icon_set", "NONE").toString ();
 
     static struct
     {
       QString name;
       QString path;
@@ -2777,21 +2793,16 @@ namespace octave
 
     qRegisterMetaType<octave_value_list> ("octave_value_list");
 
     // Even if START_GUI is false, we still set up the QApplication so
     // that we can use Qt widgets for plot windows.
 
     m_qt_app = new QApplication (m_argc, m_argv);
 
-    // set windows style for windows
-#if defined (Q_OS_WIN32)
-    m_qt_app->setStyle (QStyleFactory::create ("Windows"));
-#endif
-
     // Force left-to-right alignment (see bug #46204)
     m_qt_app->setLayoutDirection (Qt::LeftToRight);
 
     octave_link::connect_link (m_octave_qt_link);
 
     connect (m_octave_qt_link, SIGNAL (confirm_shutdown_signal (void)),
              this, SLOT (confirm_shutdown_octave (void)));
 
diff --git a/libgui/src/main-window.h b/libgui/src/main-window.h
--- a/libgui/src/main-window.h
+++ b/libgui/src/main-window.h
@@ -327,24 +327,28 @@ namespace octave
     bool focus_console_after_command (void);
 
     void new_figure_callback (void);
 
     void configure_shortcuts (void);
 
     QList<octave_dock_widget *> dock_widget_list (void);
 
+    QApplication *m_qt_app;
+
     octave_qt_link *m_octave_qt_link;
 
     workspace_model *m_workspace_model;
 
     QHash<QMenu*, QStringList> m_hash_menu_text;
 
     QString m_default_encoding;
 
+    QString m_default_style;
+
     //! Toolbar.
 
     QStatusBar *m_status_bar;
 
     //! Dock widgets.
     //!@{
     terminal_dock_widget *m_command_window;
     history_dock_widget *m_history_window;
@@ -501,16 +505,18 @@ namespace octave
     ~octave_qt_app (void);
 
     void config_translators (void);
 
     void create_main_window (void);
 
     int exec (void);
 
+    QApplication *qt_app (void) { return m_qt_app; };
+
   public slots:
 
     void handle_octave_finished (int);
 
     void confirm_shutdown_octave (void);
 
     void copy_image_to_clipboard (const QString& file, bool remove_file);
 
diff --git a/libgui/src/settings-dialog.cc b/libgui/src/settings-dialog.cc
--- a/libgui/src/settings-dialog.cc
+++ b/libgui/src/settings-dialog.cc
@@ -38,16 +38,17 @@ along with Octave; see the file COPYING.
 #include <QButtonGroup>
 #include <QDir>
 #include <QFileInfo>
 #include <QFileDialog>
 #include <QVector>
 #include <QHash>
 #include <QMessageBox>
 #include <QScrollBar>
+#include <QStyleFactory>
 #include <QTextCodec>
 
 #if defined (HAVE_QSCINTILLA)
 #  include "octave-qscintilla.h"
 #  include "octave-txt-lexer.h"
 #  include <QScrollArea>
 
 #  if defined (HAVE_QSCI_QSCILEXEROCTAVE_H)
@@ -119,16 +120,30 @@ namespace octave
     if (language == "SYSTEM")
       language = tr ("System setting");
     int selected = comboBox_language->findText (language);
     if (selected >= 0)
       comboBox_language->setCurrentIndex (selected);
     else
       comboBox_language->setCurrentIndex (0);  // System is default
 
+    // Global style
+    QStringList styles = QStyleFactory::keys();
+    combo_styles->addItems (styles);
+    combo_styles->insertItem (0, global_style.def.toString ());
+    combo_styles->insertSeparator (1);
+    QString current_style = settings->value (global_style.key, global_style.def).toString ();
+    if (current_style == global_style.def.toString ())
+      current_style = global_style.def.toString ();
+    selected = combo_styles->findText (current_style);
+    if (selected >= 0)
+      combo_styles->setCurrentIndex (selected);
+    else
+      combo_styles->setCurrentIndex (0);
+
     // icon size and theme
     QButtonGroup *icon_size_group = new QButtonGroup (this);
     icon_size_group->addButton (icon_size_small);
     icon_size_group->addButton (icon_size_normal);
     icon_size_group->addButton (icon_size_large);
     int icon_size = settings->value (global_icon_size.key, global_icon_size.def).toInt ();
     icon_size_normal->setChecked (true);  // the default
     icon_size_small->setChecked (icon_size < 0);
@@ -802,16 +817,22 @@ namespace octave
     settings->setValue ("DockWidgets/widget_icon_set", widget_icon_set);
 
     // language
     QString language = comboBox_language->currentText ();
     if (language == tr ("System setting"))
       language = "SYSTEM";
     settings->setValue ("language", language);
 
+    // style
+    QString selected_style = combo_styles->currentText ();
+    if (selected_style == global_style.def.toString ())
+      selected_style = global_style.def.toString ();
+    settings->setValue (global_style.key, selected_style);
+
     // dock widget title bar
     settings->setValue ("DockWidgets/widget_title_custom_style", cb_widget_custom_style->isChecked ());
     settings->setValue ("DockWidgets/widget_title_3d", sb_3d_title->value ());
     settings->setValue ("DockWidgets/title_bg_color", m_widget_title_bg_color->color ());
     settings->setValue ("DockWidgets/title_bg_color_active", m_widget_title_bg_color_active->color ());
     settings->setValue ("DockWidgets/title_fg_color", m_widget_title_fg_color->color ());
     settings->setValue ("DockWidgets/title_fg_color_active", m_widget_title_fg_color_active->color ());
 
diff --git a/libgui/src/settings-dialog.ui b/libgui/src/settings-dialog.ui
--- a/libgui/src/settings-dialog.ui
+++ b/libgui/src/settings-dialog.ui
@@ -91,31 +91,167 @@
                       <width>40</width>
                       <height>20</height>
                      </size>
                     </property>
                    </spacer>
                   </item>
                  </layout>
                 </item>
-                <item row="10" column="0">
-                 <widget class="QCheckBox" name="cb_prompt_to_exit">
+                <item row="7" column="0">
+                 <widget class="QLabel" name="label_15">
                   <property name="text">
-                   <string>Confirm before exiting</string>
+                   <string>Dock widget title bar</string>
+                  </property>
+                  <property name="alignment">
+                   <set>Qt::AlignLeading|Qt::AlignLeft|Qt::AlignTop</set>
+                  </property>
+                 </widget>
+                </item>
+                <item row="8" column="0">
+                 <widget class="QCheckBox" name="cb_use_native_file_dialogs">
+                  <property name="text">
+                   <string>Use native file dialogs</string>
+                  </property>
+                  <property name="checked">
+                   <bool>true</bool>
                   </property>
                  </widget>
                 </item>
-                <item row="1" column="0">
-                 <widget class="QLabel" name="label_2">
+                <item row="3" column="1">
+                 <layout class="QVBoxLayout" name="verticalLayout_5">
+                  <item>
+                   <layout class="QHBoxLayout" name="horizontalLayout_6">
+                    <item>
+                     <widget class="QRadioButton" name="icon_size_small">
+                      <property name="text">
+                       <string>Small</string>
+                      </property>
+                     </widget>
+                    </item>
+                    <item>
+                     <widget class="QRadioButton" name="icon_size_normal">
+                      <property name="text">
+                       <string>Normal</string>
+                      </property>
+                      <property name="checked">
+                       <bool>true</bool>
+                      </property>
+                     </widget>
+                    </item>
+                    <item>
+                     <widget class="QRadioButton" name="icon_size_large">
+                      <property name="text">
+                       <string>Large</string>
+                      </property>
+                     </widget>
+                    </item>
+                    <item>
+                     <spacer name="horizontalSpacer_4">
+                      <property name="orientation">
+                       <enum>Qt::Horizontal</enum>
+                      </property>
+                      <property name="sizeHint" stdset="0">
+                       <size>
+                        <width>40</width>
+                        <height>20</height>
+                       </size>
+                      </property>
+                     </spacer>
+                    </item>
+                   </layout>
+                  </item>
+                  <item>
+                   <widget class="QCheckBox" name="cb_system_icon_theme">
+                    <property name="text">
+                     <string>Use system icon theme if available (requires restart)</string>
+                    </property>
+                   </widget>
+                  </item>
+                 </layout>
+                </item>
+                <item row="9" column="0">
+                 <widget class="QCheckBox" name="cb_cursor_blinking">
                   <property name="text">
-                   <string>Language</string>
+                   <string>Cursor blinking</string>
+                  </property>
+                  <property name="checked">
+                   <bool>true</bool>
                   </property>
                  </widget>
                 </item>
-                <item row="6" column="1">
+                <item row="10" column="0">
+                 <widget class="QCheckBox" name="cb_status_bar">
+                  <property name="text">
+                   <string>Show status bar</string>
+                  </property>
+                  <property name="checked">
+                   <bool>true</bool>
+                  </property>
+                 </widget>
+                </item>
+                <item row="5" column="1">
+                 <layout class="QHBoxLayout" name="horizontalLayout_9">
+                  <item>
+                   <widget class="QRadioButton" name="general_icon_octave">
+                    <property name="text">
+                     <string>Octave logo only</string>
+                    </property>
+                    <property name="checked">
+                     <bool>true</bool>
+                    </property>
+                   </widget>
+                  </item>
+                  <item>
+                   <widget class="QRadioButton" name="general_icon_letter">
+                    <property name="text">
+                     <string>Letter icons</string>
+                    </property>
+                   </widget>
+                  </item>
+                  <item>
+                   <widget class="QRadioButton" name="general_icon_graphic">
+                    <property name="text">
+                     <string>Graphic icons</string>
+                    </property>
+                   </widget>
+                  </item>
+                  <item>
+                   <spacer name="horizontalSpacer_6">
+                    <property name="orientation">
+                     <enum>Qt::Horizontal</enum>
+                    </property>
+                    <property name="sizeHint" stdset="0">
+                     <size>
+                      <width>40</width>
+                      <height>20</height>
+                     </size>
+                    </property>
+                   </spacer>
+                  </item>
+                 </layout>
+                </item>
+                <item row="3" column="0">
+                 <widget class="QLabel" name="label_8">
+                  <property name="text">
+                   <string>Toolbar Icons</string>
+                  </property>
+                  <property name="alignment">
+                   <set>Qt::AlignLeading|Qt::AlignLeft|Qt::AlignTop</set>
+                  </property>
+                 </widget>
+                </item>
+                <item row="5" column="0">
+                 <widget class="QLabel" name="label_9">
+                  <property name="text">
+                   <string>Icon set for dock widgets</string>
+                  </property>
+                 </widget>
+                </item>
+                <item row="7" column="1">
                  <layout class="QHBoxLayout" name="horizontalLayout_4">
                   <item>
                    <layout class="QGridLayout" name="gridLayout_13">
                     <property name="rightMargin">
                      <number>0</number>
                     </property>
                     <item row="0" column="9">
                      <widget class="QLabel" name="label_bgtitle_active">
@@ -267,166 +403,57 @@
                       <width>40</width>
                       <height>20</height>
                      </size>
                     </property>
                    </spacer>
                   </item>
                  </layout>
                 </item>
-                <item row="6" column="0">
-                 <widget class="QLabel" name="label_15">
+                <item row="11" column="0">
+                 <widget class="QCheckBox" name="cb_prompt_to_exit">
                   <property name="text">
-                   <string>Dock widget title bar</string>
-                  </property>
-                  <property name="alignment">
-                   <set>Qt::AlignLeading|Qt::AlignLeft|Qt::AlignTop</set>
+                   <string>Confirm before exiting</string>
                   </property>
                  </widget>
                 </item>
-                <item row="4" column="1">
-                 <layout class="QHBoxLayout" name="horizontalLayout_9">
+                <item row="1" column="0">
+                 <widget class="QLabel" name="label_2">
+                  <property name="text">
+                   <string>Language</string>
+                  </property>
+                 </widget>
+                </item>
+                <item row="2" column="1">
+                 <layout class="QHBoxLayout" name="horizontalLayout_17">
                   <item>
-                   <widget class="QRadioButton" name="general_icon_octave">
-                    <property name="text">
-                     <string>Octave logo only</string>
-                    </property>
-                    <property name="checked">
-                     <bool>true</bool>
-                    </property>
-                   </widget>
+                   <widget class="QComboBox" name="combo_styles"/>
                   </item>
                   <item>
-                   <widget class="QRadioButton" name="general_icon_letter">
-                    <property name="text">
-                     <string>Letter icons</string>
-                    </property>
-                   </widget>
-                  </item>
-                  <item>
-                   <widget class="QRadioButton" name="general_icon_graphic">
-                    <property name="text">
-                     <string>Graphic icons</string>
-                    </property>
-                   </widget>
-                  </item>
-                  <item>
-                   <spacer name="horizontalSpacer_6">
+                   <spacer name="horizontalSpacer">
                     <property name="orientation">
                      <enum>Qt::Horizontal</enum>
                     </property>
                     <property name="sizeHint" stdset="0">
                      <size>
                       <width>40</width>
                       <height>20</height>
                      </size>
                     </property>
                    </spacer>
                   </item>
                  </layout>
                 </item>
-                <item row="4" column="0">
-                 <widget class="QLabel" name="label_9">
-                  <property name="text">
-                   <string>Icon set for dock widgets</string>
-                  </property>
-                 </widget>
-                </item>
-                <item row="9" column="0">
-                 <widget class="QCheckBox" name="cb_status_bar">
-                  <property name="text">
-                   <string>Show status bar</string>
-                  </property>
-                  <property name="checked">
-                   <bool>true</bool>
-                  </property>
-                 </widget>
-                </item>
-                <item row="8" column="0">
-                 <widget class="QCheckBox" name="cb_cursor_blinking">
+                <item row="2" column="0">
+                 <widget class="QLabel" name="label_29">
                   <property name="text">
-                   <string>Cursor blinking</string>
-                  </property>
-                  <property name="checked">
-                   <bool>true</bool>
-                  </property>
-                 </widget>
-                </item>
-                <item row="7" column="0">
-                 <widget class="QCheckBox" name="cb_use_native_file_dialogs">
-                  <property name="text">
-                   <string>Use native file dialogs</string>
-                  </property>
-                  <property name="checked">
-                   <bool>true</bool>
-                  </property>
-                 </widget>
-                </item>
-                <item row="2" column="0">
-                 <widget class="QLabel" name="label_8">
-                  <property name="text">
-                   <string>Toolbar Icons</string>
-                  </property>
-                  <property name="alignment">
-                   <set>Qt::AlignLeading|Qt::AlignLeft|Qt::AlignTop</set>
+                   <string>Style</string>
                   </property>
                  </widget>
                 </item>
-                <item row="2" column="1">
-                 <layout class="QVBoxLayout" name="verticalLayout_5">
-                  <item>
-                   <layout class="QHBoxLayout" name="horizontalLayout_6">
-                    <item>
-                     <widget class="QRadioButton" name="icon_size_small">
-                      <property name="text">
-                       <string>Small</string>
-                      </property>
-                     </widget>
-                    </item>
-                    <item>
-                     <widget class="QRadioButton" name="icon_size_normal">
-                      <property name="text">
-                       <string>Normal</string>
-                      </property>
-                      <property name="checked">
-                       <bool>true</bool>
-                      </property>
-                     </widget>
-                    </item>
-                    <item>
-                     <widget class="QRadioButton" name="icon_size_large">
-                      <property name="text">
-                       <string>Large</string>
-                      </property>
-                     </widget>
-                    </item>
-                    <item>
-                     <spacer name="horizontalSpacer_4">
-                      <property name="orientation">
-                       <enum>Qt::Horizontal</enum>
-                      </property>
-                      <property name="sizeHint" stdset="0">
-                       <size>
-                        <width>40</width>
-                        <height>20</height>
-                       </size>
-                      </property>
-                     </spacer>
-                    </item>
-                   </layout>
-                  </item>
-                  <item>
-                   <widget class="QCheckBox" name="cb_system_icon_theme">
-                    <property name="text">
-                     <string>Use system icon theme if available (requires restart)</string>
-                    </property>
-                   </widget>
-                  </item>
-                 </layout>
-                </item>
                </layout>
               </item>
              </layout>
             </widget>
            </item>
            <item>
             <widget class="QGroupBox" name="groupBox_2">
              <property name="title">
